Add initial functional tests

dkarlovi opened the pull request:
As suggested in https://github.com/rollerworks/search/issues/179

dkarlovi commented:
@sstok is this something you had in mind?

sstok commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
+/.idea/
```
Please remove this, you can use the global .gitignore instead 👍

dkarlovi replied:
Done, thanks for the link.

sstok commented:
Thank you so much for your work so far 👍

dkarlovi commented:
@sstok can you elaborate a bit on the `registerField()`? How is that supposed to work exactly?

dkarlovi commented on tests/Functional/FunctionalElasticsearchTestCase.php:
```diff
@@ -0,0 +1,252 @@
+    protected function configureConditionGenerator(QueryConditionGenerator $conditionGenerator)
+    {
+        // TODO: mapping field set alias to Elasticsearch property
+        return;
```
@sstok this is where I'm hoping you can shed some light. :)

dkarlovi commented on tests/Functional/FunctionalElasticsearchTestCase.php:
```diff
@@ -0,0 +1,252 @@
+        $conditionGenerator = new QueryConditionGenerator($condition);
+        $this->configureConditionGenerator($conditionGenerator);
+
+        // TODO: where do I get these from?
```
@sstok also here, don't quite understand how you imagined for it to work.

sstok commented:
The `registerField` method works similar to the Doctrine DBAL/ORM implementation 😃
You should be able to map a search field to one or more fields, but ElasticSearch actually is much more advanced in this topic as (IIRC) you can create an index with combining fields.

It's a while since I worked on the ElasticSearch implementation 😅

dkarlovi commented:
@sstok sure, we can work on advanced mappings later, for now let's try and get **anything** working. :)

For example, how did you imagine we map `id: 1` to `/invoices/invoices`, property `_id` with the `registerField()`?

dkarlovi commented on .php_cs:
```diff
@@ -15,7 +15,7 @@ return PhpCsFixer\Config::create()
         '@PHP70Migration' => true,
         '@PHP71Migration' => true,
-        'array_syntax' => array('syntax' => 'short'),
+        'array_syntax' => ['syntax' => 'short'],
```
This one is ironic. :sob:

dkarlovi commented:
@sstok I've added some changes to the query generator itself:

1. `getQuery` now wraps in `query` so you can use it as is (refactored your tests to match)
2. implemented a `registerField` in some manner and refactored the hardcoded mappings in query generator and (your tests)

Can you take a look and point out any glaring errors or do you feel this might be a way to go?

dkarlovi commented:
@sstok 

I've taken the McDonald's principle and have made [a property notation format proposition](https://github.com/rollerworks/search-elasticsearch/pull/1/files#diff-4a56eb64cd11341f5e66b764d17cadc9R36). If you have any feedback on this or anything else in this PR, I'd really love to hear it. Also need a review regarding normalizers (specifically date, but others too). 

I think this has quite some potential, need to work out the details.

sstok commented on src/QueryConditionGenerator.php:
```diff
@@ -23,8 +25,10 @@
+
     private $searchCondition;
-    private $fieldSet;
+    // private $fieldSet;
```
This property is used in the code?

dkarlovi replied:
Yep, my bad. :(

sstok commented on src/QueryConditionGenerator.php:
```diff
@@ -81,40 +105,79 @@ private function processGroup(ValuesGroup $group): array
+                    // https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-range-query.html#ranges-on-dates
+                    $values = array_values($valuesBag->getSimpleValues());
+                    foreach ($values as $idx => $value) {
+                        $valuesBag->removeSimpleValue($idx);
```
Note a ValuesBag is not immutable, and the QueryConditionGenerator should not change the ValuesBag.
Better to clone this.

For the future: Usually when a Query generator needs something special is to use TypeExtensions with options like `elastic_search_structure_transformer` which receives the ValuesBag (or ValuesGroup, or both) and allows to change it's structure.

dkarlovi replied:
I'm just working on this and it's annoying as it is, especially since it's only placeholder code.

sstok replied:
I did something similar for the Doctrine DBAL integration 😋 `query conversions`.
https://github.com/rollerworks/search-doctrine-dbal/blob/master/src/Query/QueryField.php#L96

For example the https://github.com/rollerworks/search-doctrine-dbal/blob/master/src/Extension/Conversion/AgeDateConversion.php allows to search with both an age and a date.

sstok commented on src/QueryConditionGenerator.php:
```diff
@@ -81,40 +105,79 @@ private function processGroup(ValuesGroup $group): array
+                foreach ($valuesBag->get(Range::class) as $range) {
+                    $ids = array_merge($ids, range($range->getLower(), $range->getUpper()));
+                }
+                if (false === empty($ids)) {
```
No need for the explicit Boolean check, you can simple use `if (f!empty($ids)) {`
Same goes for all other code occurrences.

sstok commented on src/QueryConditionGenerator.php:
```diff
@@ -81,40 +105,79 @@ private function processGroup(ValuesGroup $group): array
 
-                $bool[$includingType][] = [$this->mappings[$fieldName]->indexName => $rangeParams];
-            }
+            if (false === $isId) {
```
This check needs to be done sooner as simple values can also apply to the id, no?

dkarlovi replied:
Yes, I'll be moving the whole thing.

dkarlovi commented on src/ValueConversion.php:
```diff
@@ -0,0 +1,29 @@
+namespace Rollerworks\Component\Search\Elasticsearch;
+
+/**
+ * Interface ValueConversion.
```
Thanks. I think I have a good idea how we can push the functionality further. Basically, special cases can go to the per-type extension which should allow, for example, exceptions for dates and IDs while keeping the basic query generator quite simple.

It'll be probably done monday. How do you plan actually implementing the client, if not using Elastica?

dkarlovi commented on src/QueryConditionGenerator.php:
```diff
@@ -239,10 +236,53 @@ private function processPatternMatchers(array $values, string $fieldName, array
+        return $converter->convertValue($value);
+    }
+
+    /**
```
TBH I'm not a big fan of the selective decision process if a docblock is necessary. It puts burden on the developer / reviewer to figure out. My path usually is to add them everywhere, it doesn't hurt readability, but it does allow for my tooling to work better.

I'll remove it here, obviously. :+1:

dkarlovi commented:
Added a query conversion concept (which should handle special cases for IDs, dates, etc).

Also, you've inverted less than / greater than in your code and tests, I had a WTF moment there for a bit. :sob: Check the latest changes and feedback if you get the chance.

BTW how do I explain to the system that any date-y field should be converted by my DateConverter? Currently, BirthdayType is broken as it's not handled.

dkarlovi commented:
I've added support for nested queries (for example, binding property `items[].label` will do a nested query on `items` and do a regular query on `items.label`), will finish compare and matchers tomorrow.

What can we do to get this merged and released tomorrow? I'd like to include a preview version of this in my app ASAP (that's the reason I've done this work in the first place). If you could help me out here, it would be great.

Thanks.

sstok commented:
Amazing work! This looks good enough to move forwards.

About the Client part, I didn't want to force a specific client implementation to use the user.
For Elastica we can pass the generated query to https://github.com/ruflin/Elastica/blob/master/lib/Elastica/Query.php 👍

For the API-Platform we need a full integration (using Elastica) to also make the mapping of Resources to Indexes/Documents work. But I'm not sure what the best approach is for this.
For limiting access we need to have some Context Listener that will (based on a custom implementation) ensures the the ElasticSearch query limits the results to what's accessible by the user.

dkarlovi commented:
> Amazing work! This looks good enough to move forwards.

\o/ Could you submit the package to Packagist, it would allow us to move forward on this.

> For the API-Platform we need a full integration (using Elastica) to also make the mapping of Resources to Indexes/Documents work. But I'm not sure what the best approach is for this.

Maybe require FOS Elastica Bundle and compiler-pass the mapping into it?

sstok commented:
I'm going to merge this one so we can move forward 👍

sstok commented:
Thank you @dkarlovi

dkarlovi commented:
Wait, I'm pushing more stuff!

sstok commented:
No problem ;) I can even give you push access to this repository as you did an amazing job and I don't have always time to review everything in detail 😄

Basically I have a few rules (_I still need to formalize these_):

* Don't use the GitHub UI for making direct commits (except for fixing small typo's) - Use pull requests with [clear commit messages](https://chris.beams.io/posts/git-commit/) and use Git rebase (don't merge) for updating your local branch, be sure to clean-up any WIP commits afterwards (I didn't do this for this pull request as it would be to much work with set-up of the master branch and all).

* Use https://github.com/park-manager/hubkit for merging pull-requests and making releases (no major releases yet!!)

* Don't make a make major release (everything for RollerworksSearch v2 is still in alpha atm). Feel free to make unstable 0.x minor/patch releases (using Hubkit), note that you need to have [git signing](https://git-scm.com/book/en/v2/Git-Tools-Signing-Your-Work) enabled for this.

When in doubt about something feel free to ask me 👍 (I will always do a review afterwards)
Again thank you every much for your amazing work and trust in RollerworksSearch 😸

dkarlovi commented:
> I can even give you push access to this repository

As said before, I really think you should consider a mono-repo for this project, the overhead of setup required for every single-repo will be quite high. Not that I'm requesting push access for that mono repo. :)

For example, I like to setup [all the infra stuff](https://github.com/dkarlovi/dockerfile-dev) to be run from a Makefile and then can just run `make` before commiting.

> clear commit messages

That's why I'd rather you undo this merge as I've been doing WIP commits which I intended to fix afterward.
